Arbitrum 2crv ready for review

pata-eth opened the pull request.

pata-eth commented:
converted this one to draft until we sort out the reward claim process for all new curve gauges.

pata-eth commented:
@poolpitako and @dudesahn, given the material changes of the commit I just pushed, we may want to have poolpi review its diff so that he can confirm he still approves the PR as it currently stands. Thank you both!

pata-eth commented:
![Screen Shot 2022-05-22 at 6 54 04 PM](https://user-images.githubusercontent.com/19500282/169719715-5ba8eddc-6cac-484c-8984-6f6ee814151c.png)

pata-eth commented:
![Screen Shot 2022-05-23 at 10 05 41 PM](https://user-images.githubusercontent.com/19500282/169934602-3245e4be-339d-48c0-aba9-717a8da30f56.png)
@poolpitako and @dudesahn, I've updated this PR in line with the feedback I received for ftm-tricrypto. Please let me know when you are ready to involve security. Thank you

dudesahn commented on brownie-config.yml:
```diff
@@ -14,11 +18,43 @@ dependencies:
-    version: 0.6.12
+    version:
     remappings:
       - "@yearnvaults=yearn/yearn-vaults@0.4.3"
```
make sure to use 0.4.3-1

pata-eth replied:
I will update. The tricrypto strats in arb and ftm that re recently deployed were using 0.4.3. Are we ok, or should we update those to 0.4.3-1 as well?

dudesahn replied:
I think 0.4.3-1 is a little nicer, all it adds is events that the yearn-watch guys found helpful. and on sidechains/L2s we really don't mind a little extra gas for events imo

dudesahn commented on contracts/StrategyCurveTwoPool.sol:
```diff
@@ -0,0 +1,383 @@
+    // Arbitrum healthcheck contract address
+    healthCheck = _contractsAddress.healthCheck;
+
+    // Our standard approvals. want = Curve LP token
```
Linting needed here and elsewhere

pata-eth replied:
yeah, sorry. I'll fix it. I noticed the original source code had tab defined with 4 spaces and in my starting setup it was defined with 2. I seems that the linter didn't like me updating to 4 spaces somewhere between commits.

dudesahn commented on contracts/StrategyCurveTwoPool.sol:
```diff
@@ -0,0 +1,383 @@
+abstract contract StrategyCurveBase is BaseStrategy {
+  using Address for address;
+
+  /* ========== STATE VARIABLES ========== */
```
no keepCRV?

pata-eth replied:
I removed this variable along with the voter, etc. as we cannot currently vote on sidechains. I'm happy to add it back if we have a reasonable expectation voting will eventually make it to arbitrum.

dudesahn replied:
I would leave keepCRV, because even if we don't necessarily vote, we probably want to keep and lock of 10% of the CRV from harvests if possible

pata-eth replied:
Absolutely. I'll get the code back.

dudesahn commented on contracts/StrategyCurveTwoPool.sol:
```diff
@@ -0,0 +1,383 @@
+  function prepareMigration(address _newStrategy) internal override {
+        // Withdraw LP tokens from the gauge. The transfer to the new strategy is done
+        // by migrate() in BaseStrategy.sol. Note that this function does not claim
+        // any pending rewards. To claim and transfer pending rewards to a new
```
Can probably delete the last sentence of this comment

pata-eth replied:
will do

dudesahn commented on contracts/StrategyCurveTwoPool.sol:
```diff
@@ -0,0 +1,383 @@
+  // Credit threshold is in want token, and will trigger a harvest if credit is large enough.
+  function setCreditThreshold(uint256 _creditThreshold)
+    external
+    onlyEmergencyAuthorized
```
Any reason for `onlyEmergencyAuthorized` vs `onlyVaultManagers`? If this is just what I had in the past, then I would say the latter makes more sense. Basically `onlyVaultManagers` is the best thing to use for stuff like this imo.

pata-eth replied:
yeah, it was there. I'll update to `onlyVaultManagers`

dudesahn commented on contracts/StrategyCurveTwoPool.sol:
```diff
@@ -0,0 +1,383 @@
+    override
+    returns (uint256)
+  {
+    return _ethAmount;
```
Can delete this

pata-eth replied:
will do

dudesahn replied:
`return _ethAmount;` can be deleted since you don't use it anywhere

dudesahn commented on contracts/StrategyCurveTwoPool.sol:
```diff
@@ -0,0 +1,383 @@
+  // deposit into the curve pool -- USDT or USDC
+  address public targetTokenAddress;
+
+  IERC20 internal usdt;
```
immutable?

pata-eth replied:
interestingly, when you need to approve an allowance in the constructor (in this case for usdt), solidity does not let you set the variable to `immutable`. At any rate, I'll go back to using hard-coded addresses.

dudesahn commented on contracts/StrategyCurveTwoPool.sol:
```diff
@@ -0,0 +1,383 @@
+  // Set optimal token to sell harvested funds for depositing to Curve.
+  // Default is USDC, but can be set to USDT as needed by strategist or governance.
+  function setTargetToken(uint256 _target) external onlyVaultManagers {
+    require(_target == 0 || _target == 1); // dev: not a valid index => use 0: USDT 1: USDC
```
could probably do require < 2 to save a bit of gas

dudesahn commented on contracts/StrategyCurveTwoPool.sol:
```diff
@@ -0,0 +1,383 @@
+        // Sell CRV if we have any
+    if (crvBalance > 0) {
+      _sell(crvBalance);
+      uint256 usdtBalance = usdt.balanceOf(address(this));
```
I would move lines 291-293 outside of the if statement and use a separate if for usdc/usdt, for instance you could get USDC/USDT donations to strategy, etc

pata-eth replied:
sounds good. By now, I've seen or heard the concept of 'donation' a few times. When would this occur? Why would someone or a contract donate their tokens to the strategy?

dudesahn replied:
so realistically it's important to at least be accounting for donations _somehow_ so that they don't wreck your strategy in case anyone malicious sent some in. But also, sometimes we might need to sweep something out, sell it, and donate profits to the strategy, and it's nice if the strategy can just recognize the underlying tokens instead of us having to convert all of the way to `want` before sending it in.

dudesahn commented on contracts/StrategyCurveTwoPool.sol:
```diff
@@ -0,0 +1,383 @@
+        // Claim and get a fresh snapshot of the strategy's CRV balance
+        _claimRewards();
+
+    uint256 crvBalance = crv.balanceOf(address(this));
```
similar observation about keepCRV as above

dudesahn commented on contracts/StrategyCurveTwoPool.sol:
```diff
@@ -0,0 +1,383 @@
+    returns (address[] memory)
+  {}
+
+  /* ========== SETTERS ========== */
```
If doing keepCRV, would want a setter for that as well

dudesahn commented on contracts/StrategyCurveTwoPool.sol:
```diff
@@ -0,0 +1,383 @@
+  }
+
+    function _claimRewards() internal {
+        gaugeFactory.mint(address(gauge));
```
With the new gauge factory, is it possible to see how much a specific user is owed for a gauge before claiming? if so, we could estimate potential rewards, which we're currently not able to with Curve strategies.

pata-eth replied:
not on-chain with the current implementation, unfortunately (see [the code here](https://github.com/curvefi/curve-xchain-factory/blob/master/contracts/implementations/ChildGauge.vy#L490)). We do use the function in a few of the tests to estimate rewards before claiming to make sure the gauge is working as expected.

dudesahn commented on contracts/StrategyCurveTwoPool.sol:
```diff
@@ -0,0 +1,383 @@
+  // Uniswap V3 router
+  ISwapRouter public immutable router;
+
+  IERC20 public crv;
```
immutable?

pata-eth replied:
https://github.com/dudesahn/StrategyCurveTemplate/pull/6#discussion_r882194227

pata-eth commented:
Thanks again, for the feedback @dudesahn ! I've fixed my linter and updated the code according to your feedback. We still have a few open conversations unresolved, but I think this is ready for @poolpitako's final review. I may submit one more change based on how we resolve our pending items. Thank you

pata-eth commented:
![Screen Shot 2022-06-07 at 4 50 08 PM](https://user-images.githubusercontent.com/19500282/172479962-abba516e-5ed9-4198-a8d6-2ea86e4018c0.png)
@dudesahn, I added keepCRV back. I think I've now taken care of all the requests. Please validate at your earliest convenience. Thank you

dudesahn commented:
https://github.com/pata-eth/StrategyCurveTemplate/blob/16c2cbd3ae25ad38b42e400202ed54df5abedc95/contracts/StrategyCurveTwoPool.sol#L222

you can delete this line since you don't call that function anywhere

dudesahn commented:
non-blocking, but in the future I would encourage you to use natspec for your comments of functions—that way they show up nicely on etherscan :)

dudesahn commented:
https://github.com/pata-eth/StrategyCurveTemplate/blob/16c2cbd3ae25ad38b42e400202ed54df5abedc95/contracts/StrategyCurveTwoPool.sol#L16

Also, I think you can delete lines 16 and 18, as these are already inherited from the baseStrategy. again, not a big deal tho.

dudesahn commented:
LGTM as my comments are non-blocking

pata-eth commented:
> non-blocking, but in the future I would encourage you to use natspec for your comments of functions/state vars—that way they show up nicely on etherscan. I did this for my most recent iteration of strategies and am pretty happy with it.

Absolutely. In my new projects I'm already incorporating natspec. So much better!

pata-eth commented:
> https://github.com/pata-eth/StrategyCurveTemplate/blob/16c2cbd3ae25ad38b42e400202ed54df5abedc95/contracts/StrategyCurveTwoPool.sol#L16
> 
> Also, I think you can delete lines 16 and 18, as these are already inherited from the baseStrategy. again, not a big deal tho.

Do you know why we can remove these lines but not lines 6 and 7?

dudesahn commented:
> 

you should be able to remove those as well :)
